DRA API: implement CEL cost limit

pohly opened the pull request:
#### What type of PR is this?

/kind feature
/kind api-change
/wg structured-parameters

#### What this PR does / why we need it:

The main purpose is to protect against denial-of-service attacks.  Scheduling
time depends a lot on unpredictable factors and expected scheduling time also
varies, so no attempt is made to limit the overall time spent on evaluating CEL
expressions per claim.

#### Which issue(s) this PR fixes:

Related-to: https://github.com/kubernetes/enhancements/issues/4381

#### Special notes for your reviewer:

Expressions that were stored before in 1.31 remain valid because they are exempt from re-validation. The limit is only applied to new or modified expressions. Therefore this is not an API break and the theoretic problem of "stored expression is no longer valid under the new rules, object fails validation" is avoided.

#### Does this PR introduce a user-facing change?
```release-note
Added enforcement of an upper cost bound for DRA evaluations of CEL. The API server and scheduler now enforce an upper bound on the cost and runtime steps required for evaluating a CEL expression.
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

```docs
- [KEP]: https://github.com/kubernetes/enhancements/issues/4381
```

/assign @cici37 
/cc @jpbetz

cici37 committed:
Apply strictCost for DRA

pohly committed:
DRA API: implement CEL cost limit

The main purpose is to protect against denial-of-service attacks.  Scheduling
time depends a lot on unpredictable factors and expected scheduling time also
varies, so no attempt is made to limit the overall time spent on evaluating CEL
expressions per claim.

pohly commented:
/wg device-management

pohly commented:
/priority important-soon
/milestone v1.32

This blocks beta graduation, which in turn is something that we want to merge early in the 1.32 cycle to unblock others who need to modify the API for new features.

pohly commented:
/retest

Image pull errors (too many requests). I hope retesting doesn't make the situation worse...

seans3 commented:
/assign @cici37 
/triage accepted

jpbetz commented on staging/src/k8s.io/api/resource/v1alpha3/types.go:
```diff
@@ -532,10 +532,34 @@ type CELDeviceSelector struct {
+	// cost of evaluating it is also limited based on the estimated number
+	// of logical steps. Validation against those limits happens only when
+	// setting an expression for the first time or when changing
+	// it. Therefore it is possible to change these limits without
+	// affecting stored expressions. Those remain valid.
```
```suggestion
	// it.  If the limits are changed in a future Kubernetes release, existing users are guaranteed that existing expressions will continue to be valid.
```

(To make it clear that there is no way for Kubernetes users to change the limits)

pohly replied:
Applied.

pohly replied:
I started to wonder about this... Suppose some stored expression passed the cost limit check in 1.32 and in 1.33 we reduce the limit. The expression is still valid (as stated here and implemented in the validation code), but the same reduced limit is also used as runtime limit with `cel.CostLimit(resourceapi.CELSelectorExpressionMaxCost)`.

Doesn't that mean that a valid expression then can fail to get evaluated at runtime?

Is that runtime cost check really providing some additional value? Either the cost estimation is correct, then the runtime check isn't needed, or the estimation is just a guess, then perhaps it shouldn't be used for validation because it gives the wrong impression of "this expression is fine"?

sftim replied:
There's enough CEL now that a writeup (blog article ?) for https://k8s.dev/ would help. My opinion anyway.

pohly replied:
> So, in practice we've been very careful not to change cost on users.

So we can increase it (with care, like considering downgrade scenarios), but not reduce it. I thought we could be more flexible here, but not if we then also apply a runtime check.

>  > Is that runtime cost check really providing some additional value?
>
> Having a way to halt execution at runtime seems pragmatic.

Doing that based on context cancellation makes sense. If evaluation times out, it tells admins that they need a faster machine or that they have to tweak those timeouts. This is entirely theoretic for DRA though, because there are no such limits at the moment.

But I am still not sure why DRA CEL should enable the additional runtime cost check. Can I use `cel.CostLimit` to disable the default cost limit checking? Then I would get the behavior described above ("existing users are guaranteed that existing expressions will continue to be valid")  plus "won't fail at runtime", which I find useful.

It gives us a bit more freedom to change this limit if that turns out to be useful.

> We typically don't say anything in the API docs about changing cost, so it might be the best to to the same here 

Agreed, it doesn't need to be in the user-facing OpenAPI spec. But I want it on the `CELSelectorExpressionMaxCost` constant as guidance for what changes are and what aren't allowed.

pohly replied:
> Agreed, it doesn't need to be in the user-facing OpenAPI spec. But I want it on the CELSelectorExpressionMaxCost constant as guidance for what changes are and what aren't allowed.

Changed. I'll squash that when we are done with the review.

/hold

> But I am still not sure why DRA CEL should enable the additional runtime cost check.

I went ahead and disabled that runtime cost check in an additional commit because that makes more sense to me. Please ~~yell at me~~ politely point out why I am mistaken here :grin: 

I also modified one unit test case such that it triggers the runtime cost check before that commit and no longer triggers with it.

pohly replied:
I added a benchmark because I was curious whether I would see a difference when I hack out the cost tracking at runtime. I got some (according to benchstat) statistically relevant improvements, but not enough to make further changes.

The important aspect is that clients should trust the cost estimate in the apiserver validation, which is in this PR now.

jpbetz replied:
I'd prefer we keep runtime cost in place.

My perspective comes from having worked on the internals of CEL to enable the cost systems, and I trust runtime cost more than I trust estimated cost.  The logic is simpler and more "obviously correct" than estimated cost.  I'm worried something might slip through if we don't have a runtime limit.

cc @cici37, @liggitt

pohly replied:
So we end up with the "validation accepted a CEL expression, runtime disagrees and aborts" error scenario.

I am less worried about evaluation taking longer than expected for DRA because scheduling times are already pretty fluid, depending on how complex the problem is. But I'm okay with keeping the runtime limit. I can't update the code right away (on a short vacation this week), but will do so soon.

The comment about the cost limit then becomes something along the lines of "the limit can be changed because stored expressions remain valid" but with the caveat that "clients, in particular the scheduler, must be recompiled to use the same modified limit as the apiserver", which affects version skew scenarios.

jpbetz replied:
| So we end up with the "validation accepted a CEL expression, runtime disagrees and aborts" error scenario.

True, but only if estimated cost is wrong (that is, it's bugged), since the guarantee is min(estimated cost) <= runtime cost <= max(estimated cost) when everything is working as intended.

| I can't update the code right away (on a short vacation this week), but will do so soon.

Sounds good. 

| The comment about the cost limit then becomes something along the lines of "the limit can be changed because stored expressions remain valid" but with the caveat that "clients, in particular the scheduler, must be recompiled to use the same modified limit as the apiserver", which affects version skew scenarios.

Another option would be to just not say anything changing limits here.  If we do change them, we can document the situation at that time.  I think it's great you've though the situation through though... having the CEL expressions run in the scheduler does complicate this situation.

pohly replied:
I've restored the runtime limit in my recent push: https://github.com/kubernetes/kubernetes/compare/6ad744da6d14bc149930eb9ad0a715aad0c2dd40..d53cb79cec11094099205a8c114a416b430a6c2c

> Another option would be to just not say anything changing limits here.

These constants don't show up in the Kubernetes API docs for end users. They are there for my fellow developers who may have to continue maintaining DRA when you or I are not available to answer questions. So capturing our discussion now in the code IMHO is useful.

> having the CEL expressions run in the scheduler does complicate this situation

Indeed. I would still prefer to not do that, but let's proceed with it :sweat_smile:

pohly replied:
> Changed. I'll squash that when we are done with the review.
> 
> /hold

Commit history was already cleaned before the recent LGTMs.

/hold cancel

jpbetz commented on staging/src/k8s.io/api/resource/v1alpha3/types.go:
```diff
@@ -532,10 +532,34 @@ type CELDeviceSelector struct {
+//
+// According to
+// https://github.com/kubernetes/kubernetes/blob/4aeaf1e99e82da8334c0d6dddd848a194cd44b4f/staging/src/k8s.io/apiserver/pkg/apis/cel/config.go#L20-L22,
+// this gives roughly 0.1 second for each expression evaluation.
```
This sounds about right (I've been using 70ns as my estimate for a single cost unit, which gives 0.07s).  So long as your use case supports this cost, it should be safe.

jpbetz reviewed (commented):
CEL usage LGTM.  I added to minor comments.

sftim commented:
Changelog suggestion
```diff
-DRA: The apiserver and scheduler now enforce an upper bound on the cost and runtime steps required for evaluating a CEL expression.
+Added enforcement of an upper cost bound for DRA evaluations of CEL. The API server and scheduler now enforce an upper bound on the cost and runtime steps required for evaluating a CEL expression.
```

cici37 commented on staging/src/k8s.io/dynamic-resource-allocation/cel/compile.go:
```diff
@@ -122,18 +130,35 @@ func (c compiler) CompileCELExpression(expression string, envType environment.Ty
+		//
+		// This call here should override that. In practice it shouldn't
+		// matter because the limits are the same.
+		cel.CostLimit(resourceapi.CELSelectorExpressionMaxCost),
```
Why we want to overwrite it if the limit is the same?
FYI the limit is not set only for VAP but currently for all cel expression used in Kubernetes(VAP, CRD validation rules, webhook, authz, etc)

pohly replied:
Because it's not guaranteed to be the same. We might decide to use a different limit for DRA.

Different usages for CEL have different timing requirements and thus should decide individually about their cost limits. I find it odd that the shared environment package sets a limit that is defined for VAP.

cici37 replied:
Wouldn't it be better to reuse the same limit until we decided to have a different limit for DRA?

pohly replied:
Why? They are conceptually different. One is for VAP, the other for DRA. Reusing the VAP constant would give the wrong impression that they must be the same.

pohly replied:
To make it more obvious that the code is not required to use that limit in particular, I'm now using it merely as default for an `Options` field. It's still a constant in the DRA API.

cici37 commented on staging/src/k8s.io/dynamic-resource-allocation/cel/compile.go:
```diff
@@ -122,18 +130,35 @@ func (c compiler) CompileCELExpression(expression string, envType environment.Ty
+		// This call here should override that. In practice it shouldn't
+		// matter because the limits are the same.
+		cel.CostLimit(resourceapi.CELSelectorExpressionMaxCost),
+		cel.CostTracking(estimator),
```
It seems duplicated here since we have it set in https://github.com/kubernetes/kubernetes/blob/e673417529d9d2d2dd04f28f8437cbe1c403dfbe/staging/src/k8s.io/apiserver/pkg/cel/environment/base.go#L190 when true /* strictCost */) passed.

pohly replied:
That wasn't obvious. I agree, this can be removed.

cici37 commented on staging/src/k8s.io/api/resource/v1alpha3/types.go:
```diff
@@ -532,10 +532,34 @@ type CELDeviceSelector struct {
+// https://github.com/kubernetes/kubernetes/blob/4aeaf1e99e82da8334c0d6dddd848a194cd44b4f/staging/src/k8s.io/apiserver/pkg/apis/cel/config.go#L20-L22,
+// this gives roughly 0.1 second for each expression evaluation.
+// However, this depends on how fast the machine is.
+const CELSelectorExpressionMaxCost = 1000000
```
Curious why the const added here. Could it use the same value as in https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/apis/cel/config.go ?

pohly replied:
See reply in https://github.com/kubernetes/kubernetes/pull/128101#discussion_r1803540262.

cici37 commented on staging/src/k8s.io/api/resource/v1alpha3/types.go:
```diff
@@ -532,10 +532,34 @@ type CELDeviceSelector struct {
+const CELSelectorExpressionMaxCost = 1000000
+
+// CELSelectorExpressionMaxLength is the maximum length of a CEL selector expression string.
+const CELSelectorExpressionMaxLength = 10 * 1024
```
Could we add all the CEL related const into https://github.com/kubernetes/kubernetes/blob/master/staging/src/k8s.io/apiserver/pkg/apis/cel/config.go ?

pohly replied:
See discussion in https://github.com/kubernetes/kubernetes/issues/127889: I think the API package should define its own constants because they are part of that API, instead of implicitly inheriting those constants from a shared helper package.

cici37 commented on staging/src/k8s.io/dynamic-resource-allocation/cel/compile.go:
```diff
@@ -122,18 +130,35 @@ func (c compiler) CompileCELExpression(expression string, envType environment.Ty
+		return compilationResult
+	}
+
+	compilationResult.MaxCost = costEst.Max
```
If the cost limit is a const, why don't check it here directly?

pohly replied:
I'm using the same approach as for VAP here: the validation code generates a validation error when the limit is exceeded.

This separates unrelated concerns:
- how to compile (here)
- what limits to impose (API)

pohly replied:
... although I must admit that the separation is not perfect because of the `cel.CostLimit(resourceapi.CELSelectorExpressionMaxCost)` call. I'm still not sure whether such a runtime limit makes sense (https://github.com/kubernetes/kubernetes/pull/128101#discussion_r1803421102).

pohly replied:
[Removing the runtime cost checking](https://github.com/kubernetes/kubernetes/pull/128101#discussion_r1805053152) removed this usage of `resourceapi.CELSelectorExpressionMaxCost` in `compile.go`.

pohly replied:
... and now it's back as a default for some caller-specified `Options` struct.

pohly committed:
DRA CEL: add test case for runtime cost limit check

At the moment, the cost also gets checked at runtime. This test case
ensures that this check is really active.

pohly committed:
DRA CEL: fix error checking in unit test

Not getting an expected error wasn't checked.

pohly committed:
DRA CEL: avoid redundant cel.CostTracking

It's already called by the base environment.

pohly committed:
DRA CEL: disable runtime cost check

In DRA, the cost check is done only at validation time.  At runtime, any
expression that passed validation gets executed without interrupting it. The
advantage is that it becomes easier to change the limit because stored
expression do not suddenly fail after an up- or downgrade. The limit could
even become a configuration parameter of the apiserver because that is the only
place where the limit gets checked

pohly committed:
DRA CEL: log actual cost

This may be useful for validating the cost estimate.

pohly committed:
DRA CEL: add benchmark

Expression evaluation in all scenarios gets benchmarked where compilation
works. A pending optimization in another PR caches compiled expressions, so the
time for compilation will become less important. What matters is the actual
evaluation.

knabben commented:
Hello  @pohly @jpbetz @sftim 
A reminder: The code freeze is starting 02:00 UTC Friday November 8th 2024 (about 3 weeks from now), and while there is still time, we want to ensure that each PR has a chance to be merged on time.
As the PR is tagged for 1.32, is it still planned for this release?

pohly committed:
DRA cel: enforce runtime limit by default again

As pointed out during code review, the CEL cost estimates are not considered
perfectly reliable. Therefore it is better to also do runtime checks.

Some downstream users might decide to allow CEL expressions to run
longer. Therefore the cost limit is now part of an Options struct.
kube-scheduler uses the default cost limit defined in the resource.k8s.io API,
which is the same cost limit that also the apiserver uses during validation.

jpbetz commented:
/lgtm
For usage of CEL from my prospective.  Thanks @pohly!

jpbetz commented:
/assign @thockin 
This is ready for API review

cici37 commented:
My questions are mainly around:
* Do we wanna the cel related const added in type.go? We normally don't put this kind of info inside type.go. This might need to confirm with an API approver. We currently have all cel related config/const put inside https://github.com/kubernetes/kubernetes/blob/d53cb79cec11094099205a8c114a416b430a6c2c/staging/src/k8s.io/apiserver/pkg/apis/cel/config.go
* If the cel limit used is the same number as default, why overwrite it for DRA and why make it configurable?
Note that the limit set should not be modified unless it is really needed and modifying the limit later would be considered as API change and break backward compatible which should be avoided especially for after beta features.
* There is no overall cost budget set per resource which we normally enforce for other similar resources using cel expression. E.g. for all expressions set in one ResourceClaim, we might wanna have a overall limit set for resource constraints consideration.(we normally put a number 10 times than per expression limit)

pohly commented:
> Do we wanna the cel related const added in type.go? We normally don't put this kind of info inside type.go. This might need to confirm with an API approver. 

IMHO it would be better if all APIs defined their constants. It's much easier to review and use that way. We have done the same for all other DRA constants and do the same here for the CEL cost limit.

> If the cel limit used is the same number as default, why overwrite it for DRA

Because it's conceptually a different constant. It just happens to have the same value. I don't want to write code that is based on the assumption that the value is the same, because we might still decide to change the DRA limit later.

> why make it configurable

Because I want to this staging repo Go API to be flexible. API breaks in staging repos can be very annoying. Better design the API now so that the code can be used in different scenarios (with different limits) and can be extended later (adding new fields is not an API break).

> There is no overall cost budget set per resource which we normally enforce for other similar resources using cel expression.

DRA has different requirements than VAP. @klueska, can you confirm that we don't want to enforce some fixed overall budget for all claims and requests? The intended effect is that scheduling a pod with complex hardware requirements is allowed to take longer.

cici37 commented:
/lgtm to the cel cost implementation.

klueska commented:
/lgtm

I agree with @pohly that it's better to define our own constants (even if they currently have the same value) so that we can change them later if required. It also makes it more self-contained (and readable) since the names of the constants used in the DRA packages are clearly related to DRA.

thockin commented:
I thought this would be more controversial, but it's a good analysis. I'm not sure that most users will really comprehend it, but if this is something people are tripping over on a regular basis, there's probably something wrong in the design overall

thockin commented:
/approve
